File: GatekeeperHelper/AuthorizationTool.m
//  AuthorizationTool.m

#import "AuthorizationTool.h"

@implementation AuthorizationTool

+ (BOOL)runCommand:(NSString *)command error:(NSString * _Nullable * _Nullable)errorOut {
    // 将命令放到 /bin/bash -lc "..."，并处理引号转义，确保 PATH 与登录 Shell 一致
    NSString *escaped = [command stringByReplacingOccurrencesOfString:@"\"" withString:@"\\\""];
    NSString *bash = [NSString stringWithFormat:@"/bin/bash -lc \\\"%@\\\"", escaped];

    NSString *scriptSource = [NSString stringWithFormat:
                              @"do shell script \"%@\" with administrator privileges",
                              bash];

    NSAppleScript *appleScript = [[NSAppleScript alloc] initWithSource:scriptSource];
    NSDictionary *errInfo = nil;
    NSAppleEventDescriptor *result = [appleScript executeAndReturnError:&errInfo];

    if (!result) {
        if (errorOut) {
            // AppleScript 常见错误码：
            // -128 用户取消； 1.. 任意非零为命令本身非 0 退出码
            NSString *msg = errInfo[NSLocalizedDescriptionKey];
            if (!msg) msg = errInfo[@"NSAppleScriptErrorMessage"];
            if (!msg) msg = [NSString stringWithFormat:@"执行失败（错误码：%@）",
                             errInfo[@"NSAppleScriptErrorNumber"] ?: @(-1)];
            *errorOut = msg;
        }
        return NO;
    }

    // 走到这里说明命令实际以 0 退出（成功）
    return YES;
}

@end

File: GatekeeperHelper/AuthorizationTool.swift
import Foundation

@objc class AuthorizationTool: NSObject {
    @objc static func runCommand(_ command: String, error errorOut: UnsafeMutablePointer<NSString?>?) -> Bool {
        // 构造完整 AppleScript 命令
        let fullCommand = "do shell script \"\(command.replacingOccurrences(of: "\"", with: "\\\""))\" with administrator privileges"

        var error: NSDictionary?
        if let script = NSAppleScript(source: fullCommand) {
            let result = script.executeAndReturnError(&error)
            if error == nil {
                return true
            } else {
                if let errorMessage = error?[NSAppleScript.errorBriefMessage] as? NSString {
                    errorOut?.pointee = errorMessage
                } else {
                    errorOut?.pointee = "未知错误"
                }
                return false
            }
        } else {
            errorOut?.pointee = "无法创建 AppleScript 实例"
            return false
        }
    }
}
